src/flasher_spi.c

```diff
@@ -537,6 +537,251 @@ NETX_CONSOLEAPP_RESULT_T spi_flash(const FLASHER_SPI_FLASH_T *ptFlashDescription
+ *            to be in sorted order with the smallest erase operation being at element 0
+ * 
+ * @param ptFlashDescription [in]  Device information returned by spi_detect.
+ * @param ulStartAdr         [in]  Start offset of the first erase block to be erased.
+ * @param ulEndAdr           [in]  End offset of the last erase block to be erased (offset of the last byte + 1).
+ * 
+ * @return
+ * - NETX_CONSOLEAPP_RESULT_OK: success, the memory has been erased.
+ * - NETX_CONSOLEAPP_RESULT_ERROR: An error has occurred.
+ */
+NETX_CONSOLEAPP_RESULT_T spi_smart_erase(const FLASHER_SPI_FLASH_T *ptFlashDescription, const unsigned long ulStartAdr, const unsigned long ulEndAdr)
+{
+	#if ASIC_TYP==ASIC_TYP_NETIOL
+	// Not supported due to memory constraints, fall back to normal erase
+	return spi_erase(ptFlashDescription, ulStartAdr, ulEndAdr);
+
+	// Supress -Wunused errors, will never be called
+	unsigned int tmp;
+	bitmapWriteBit(0,0,&tmp);
+	bitmapReadBit(0,&tmp);
+	sectorIsEmpty(0,(unsigned char*)&tmp);
+	#else
+	/* Be Pessimistic */
+	NETX_CONSOLEAPP_RESULT_T retVal = NETX_CONSOLEAPP_RESULT_ERROR;
+	const SPIFLASH_ATTRIBUTES_T attributes = ptFlashDescription->tAttributes;
+	const FLASHER_SPI_ERASE_T *eraseTypes = ptFlashDescription->tSpiErase;
+	
+	/* Nr of valid entries */
+	unsigned int nrEraseOps = ptFlashDescription->usNrEraseOperations;
+
+	/* Get sector size, set a cap if it exceeds the read buffer size */
+	unsigned long sectorSizeBytes = attributes.ulSectorPages * attributes.ulPageSize;
+	if(sectorSizeBytes > SPI_BUFFER_SIZE)
+	{
+		sectorSizeBytes = SPI_BUFFER_SIZE;
+	}else if(sectorSizeBytes == 0){
+		uprintf("! Could not get sector size. Page size: %d, Sector pages: %d\n", attributes.ulPageSize, attributes.ulSectorPages);
+		return NETX_CONSOLEAPP_RESULT_ERROR;
+	}
+
+	/* This assumes the erase entries are sorted (which they should be) */
+	unsigned long maxEraseSize = eraseTypes[nrEraseOps-1].Size;
+
+	/* Maximum number of chunks we consider at once, limited by the largest erase instruction */
+	unsigned long nrSectorsInLargestErase = maxEraseSize/sectorSizeBytes;
+
+	/* Number of bytes we need in the sector bitmap (+1 to account for rounding) */
+	unsigned int bitPerMapEntry = (sizeof(unsigned int)*8);
+	unsigned int bitmapSize = nrSectorsInLargestErase / bitPerMapEntry + 1;
+	
+	/* Bitmap to store the erase status by sector */
+	#define SPI_ERASE_BITMAP_ENTRY_CNT 32 /* This should be more than enough, allows erase commands of up to 1024 sectors */
+	unsigned int sectorBitmap[SPI_ERASE_BITMAP_ENTRY_CNT];
+	for (unsigned int i = 0; i < SPI_ERASE_BITMAP_ENTRY_CNT; i++)
+	{
+		sectorBitmap[i] = 0;
+	}
+	
+	if(bitmapSize > SPI_ERASE_BITMAP_ENTRY_CNT){
+		uprintf("! Remembering too many memory sections required for largest erase command, falling back to normal erase\n");
+		retVal = spi_erase(ptFlashDescription, ulStartAdr, ulEndAdr);
+		return retVal;
+	}
```

Comment:
If the largest erase block is too large to handle, we could still use the smaller block sizes.
I'd check the erase block sizes and remove any block sizes that are too small or too large, either  while reading the erase commands from SFDP or later in smart_erase.